geomstats/_mesh.py

```diff
@@ -0,0 +1,44 @@
+"""Mesh-related data structures."""
+
+import geomstats.backend as gs
+
+
+class Surface:
```

Comment:
I understand why we need the faces-vertices tuple in the computations... though could we avoid creating another class?

Because:
- we have a class explosion in Geomstats: so many classes of different types... the architecture diagrams start to be messy. I would be in favor of only using classes when we leverage inheritance.
- I remember other packages (like scikit-timeseries) who made the choice of encompassing their data into objects, and had many problems for compatibility with other python packages afterwards, possibly regretting this choice?
- doesn't it break automatic differentiation to have data encompassed into objects?

Reply:
`Surface` has two purposes:

1. it allows to keep our api simple: e.g. `log(point, base_point)`, `geodesic(initial_point, end_point=end_point)`

2. it caches information to avoid repeated computation of surface information (e.g. faces normals, areas, and centroids); this is especially relevant when computing varifold loss of a point against the same point (`end_point`) during optimization


**More on the first point**

Some packages choose to use two arrays to represent a surface mesh (`vertices`, `faces`). We can't do it without changing our API: e.g. `log(vertices, faces, base_vertices, base_faces)`. Besides being annoying, we couldn't use the methods for e.g. `QuotientMetric` as is currently the case (notice the only method we specialize is `discrete_geodesic_bvp`).

We could choose to represent points as a `tuple` instead of a class. `(vertices, faces)`. This keeps our API valid, though 1) we do not have anymore the possibility of caching (important for second point); 2) any time we need information on a mesh we need to have a function taking `vertices` and `faces`, which we cannot override (right now we do `mesh.face_centroids`; we could provide a different method to compute centroids by defining a different `Surface` object); 3) we could only handle batches with lists `list[Surface]`, as tuples would be reserved for the surface representation (with current implementation, batch can be `list[Surface]` or `tuple[Surface]`). 


**Other comments**

Compatibility should not be an issue, as this object is very simple. Anyone could even use an encapsulated version of their surface representation by creating a new mesh object with the following attributes (`vertices`, `faces`, `face_centroids`, `face_normals`, `face_areas`). In this case, they would be the ones taking advantage of inheritance.
It can be annoying, but I can't see any better alternative in this case.

No problem with automatic differentiation (notice we use it to solve the optimization problem).